R/validate-policy.R

```diff
@@ -1,26 +1,35 @@
 #' Validate Policy File
 #' 
 #' Flattens and validates policy file
-#'  @param policy YAML policy file
-#'  @param gs Gradescope data
-#'  @importFrom purrr map discard
-#'  @export
+#' @param policy YAML policy file
+#' @param gs Gradescope data
+#' @importFrom purrr map discard
+#' @export
 validate_policy <- function(policy, gs){
   policy <- flatten_policy(policy)
-  # drop categories with unavailable assignments
-  categories <- map(policy$categories, "category") |> unlist()
-  # assignments is a vector that also includes category names
-  assignments <- c(get_assignments(gs), categories)
-  # this is because some categories have only nested categories as their assignments
-  # these categories should not be dropped
-  policy$categories <- map(policy$categories, function(cat){
-    cat$assignments <- cat$assignments[cat$assignments %in% assignments]
-    if (length(cat$assignments) == 0){
-      return (NULL)
-    }
-    return (cat)
-  }) |>
-    purrr::discard(is.null)
+  prev_length <- 0
+  current_length <- length(policy$categories)
+  #keep dropping until no more drops necessary
+  while (prev_length != current_length){
+    prev_length <- length(policy$categories)
+    categories <- map(policy$categories, "category") |> unlist()
+    assignments <- c(get_assignments(gs), categories)
+    policy$categories <- map(policy$categories, function(cat){
+      # drop categories with unavailable assignments/nested categories
+      cat$assignments <- cat$assignments[cat$assignments %in% assignments]
+      if (length(cat$assignments) == 0){
```

Comment:
We should experiment with this, but have you tried if it is in fact necessary to check if length(cat$assignments) == 0 to return NULL? I think is may be the default behavior of boolean indexing with all False values anyway. Removing this could significantly help speed, since checking the length of assignments could take some time for large categories and if statements also take several clock cycles to run (rather slow). Since this savings would happen for each category for each iteration, this could be a big time savings.